src/v/cluster/partition.cc

```diff
@@ -613,6 +616,23 @@ partition::transfer_leadership(std::optional<model::node_id> target) {
     } else if (_tm_stm) {
         stm_prepare_lock = co_await _tm_stm->prepare_transfer_leadership();
     }
+
+    std::optional<ss::deferred_action<std::function<void()>>> complete_archiver;
```

Comment:
why the sequence is inverted? it should be possible to call that function after `prepare_transfer_leadership` 
is it to guarantee that it's called in case if `prepare_transfer_leadershipt` throws?

Reply:
>is it to guarantee that it's called in case if prepare_transfer_leadershipt or do_transfer_leadership throws?

Yes, exactly.  complete_transfer_leadership is safe to call any time, so we set up the `deferred_action` before trying to do anything, in case something throws.